tide-whisperer.go

```diff
@@ -149,45 +156,52 @@ func main() {
 		mongoSession := session.Copy()
 		defer mongoSession.Close()
 
-		// the Sort here has to use the same fields in the same order
-		// as the index, or the index won't apply
-		iter := mongoSession.DB("").C(deviceDataCollection).
-			Find(bson.M{"$or": []bson.M{
-			bson.M{"groupId": groupId},
-			bson.M{"_groupId": groupId, "_active": true}}}).
-			Sort("groupId", "_groupId", "time").
-			Iter()
-
-		failureReturnCode := 404
-		first := false
-		var result deviceData //generic type as device data can be comprised of many things
-		for iter.Next(&result) {
-			result.filterUnwantedFields()
-
-			bytes, err := json.Marshal(result)
-			if err != nil {
-				log.Print("Failed to marshall event: ", result, err)
-			} else {
-				if !first {
-					res.Header().Add("content-type", "application/json")
-					res.Write([]byte("["))
-					first = true
-				} else {
-					res.Write([]byte(",\n"))
-				}
-				res.Write(bytes)
-			}
+		//select this data
+		groupDataQuery := bson.M{"$or": []bson.M{bson.M{"groupId": groupId}, bson.M{"_groupId": groupId, "_active": true}}}
+		//don't return these fields
+		removeFieldsForReturn := bson.M{"_id": 0, "_groupId": 0, "_version": 0, "_active": 0, "createdTime": 0, "modifiedTime": 0, "groupId": 0}
+
+		var results []interface{}
+
+		log.Println(DATA_API_PREFIX, fmt.Sprintf("mongo query [%#v]", groupDataQuery))
+
+		startQueryTime := time.Now()
+
+		//return un-ordered (i.e. the order isn't guaranteed by mongo)
+		err := mongoSession.DB("").C(deviceDataCollection).
+			Find(groupDataQuery).
+			Select(removeFieldsForReturn).
+			All(&results)
+
+		if err != nil {
+			log.Println(DATA_API_PREFIX, fmt.Sprintf("mongo query took [%.5f]secs but failed with error [%s] ", time.Now().Sub(startQueryTime).Seconds(), err.Error()))
+			log.Println(DATA_API_PREFIX, error_running_query, err.Error())
+			http.Error(res, error_running_query, http.StatusInternalServerError)
+			return
 		}
-		if err := iter.Close(); err != nil {
-			log.Print("Iterator ended with an error: ", err)
-			failureReturnCode = 500
+
+		log.Println(DATA_API_PREFIX, fmt.Sprintf("mongo query took [%.5f]secs and returned [%d] records", time.Now().Sub(startQueryTime).Seconds(), len(results)))
+
+		if len(results) == 0 {
+			log.Println(DATA_API_PREFIX, fmt.Sprintf("completed in [%.5f]secs", time.Now().Sub(start).Seconds()))
+			log.Println(DATA_API_PREFIX, query_no_data)
+			res.Header().Add("content-type", "application/json")
+			res.WriteHeader(http.StatusNotFound)
```

Comment:
Should this use the http.Error(res, ...) mechanism or perhaps http.NotFound(...)?